src/wasmFunctions.ts

```diff
@@ -0,0 +1,81 @@
+import { safeStringFromCharCode } from "./utils/utf8";
+
+// WASM=never - disable WASM functions
+// WASM=force - force to use WASM functions
+const WASM: string = process.env.MSGPACK_WASM || process.env.WASM || "";
+export const NO_WASM = WASM === "never";
+export const FORCE_WASM = WASM === "force";
+
+type pointer = number;
+
+// WM stands for WasmModule, but not the WebAssembly.Module instance but the WebAssembly.Instance.prototype.exports
+const wm: any = (() => {
+  if (NO_WASM) {
+    return null;
+  }
+
+  try {
+    return require("../dist/wasm/msgpack.wasm.js");
+  } catch (e) {
+    if (FORCE_WASM) {
+      throw e;
+    }
+    return null;
+  }
+})();
+
+export const WASM_AVAILABLE = !!wm;
+
+// A hint of when to use WASM ver.
+export const WASM_STR_THRESHOLD = FORCE_WASM ? 0 : 1024;
+
+function setMemoryU8(destPtr: pointer, src: Uint8Array, size: number) {
+  const destView = new Uint8Array(wm.memory.buffer, destPtr, size);
+  destView.set(src);
+}
+
+function setMemoryStr(destPtr: pointer, destByteLength: number, str: string, strLength: number) {
+  const inputView = new DataView(wm.memory.buffer, destPtr, destByteLength);
+  for (let i = 0; i < strLength; i++) {
+    inputView.setUint16(i * 2, str.charCodeAt(i));
+  }
+}
+
+/**
+ * It encodes string to MessagePack str family (headByte/size + utf8 bytes).
+ * @returns The whole byte length including headByte/size.
+ */
+export function utf8EncodeWasm(str: string, output: Uint8Array): number {
+  const strLength = str.length;
+  const inputByteLength = strLength * 2;
+  const inputU16BePtr: pointer = wm.malloc(inputByteLength);
+  setMemoryStr(inputU16BePtr, inputByteLength, str, strLength);
+
+  const maxOutputHeaderSize = 1 + 4; // headByte + u32
+  const outputPtr: pointer = wm.malloc(maxOutputHeaderSize + strLength * 4);
```

Comment:
I guess better allocate this inside `utf8EncodeUint16Array` on AS side. This reduce overhead for one js <-> wasm call

Reply:
Tried to reduce a `malloc()` call there but the performance score didn't get better.

```patch
diff --git a/assembly/utf8EncodeUint16Array.ts b/assembly/utf8EncodeUint16Array.ts
index 195991f..536311c 100644
--- a/assembly/utf8EncodeUint16Array.ts
+++ b/assembly/utf8EncodeUint16Array.ts
@@ -29,9 +29,15 @@ function storeStringHeader(outputPtr: usize, utf8ByteLength: usize): usize {
 // outputPtr: u8*
 // inputPtr: u16*
 // It adds MessagePack str head bytes to the output
-export function utf8EncodeUint16Array(outputPtr: usize, inputPtr: usize, inputLength: usize): usize {
+export function utf8EncodeUint16Array(inputPtr: usize, inputLength: usize): usize {
+  const maxOutputHeaderSize = sizeof<u8>() + sizeof<u32>();
+
+  // outputPtr: [u32 outputBufferSize][outputBuffer]
+  let outputPtr = memory.allocate(maxOutputHeaderSize + inputLength * 4);
+  let outputPtrStart = outputPtr + sizeof<u32>();
+
   let utf8ByteLength = utf8CountUint16Array(inputPtr, inputLength);
-  let strHeaderOffset = storeStringHeader(outputPtr, utf8ByteLength);
+  let strHeaderOffset = storeStringHeader(outputPtrStart, utf8ByteLength);
 
   const u16s = sizeof<u16>();
   let inputOffset = inputPtr;
@@ -76,5 +82,7 @@ export function utf8EncodeUint16Array(outputPtr: usize, inputPtr: usize, inputLe
     store<u8>(outputOffset++, (value & 0x3f) | 0x80);
   }
 
-  return outputOffset - outputPtr;
+  let outputByteLength = outputOffset - outputPtrStart;
+  storeUint32BE(outputPtr, outputByteLength);
+  return outputPtr;
 }
diff --git a/src/wasmFunctions.ts b/src/wasmFunctions.ts
index 5a3c44a..3a6ab4b 100644
--- a/src/wasmFunctions.ts
+++ b/src/wasmFunctions.ts
@@ -54,9 +54,11 @@ export function utf8EncodeWasm(str: string, output: Uint8Array): number {
   const maxOutputHeaderSize = 1 + 4; // headByte + u32
   const outputPtr: pointer = wm.malloc(maxOutputHeaderSize + strLength * 4);
   try {
-    const outputLength = wm.utf8EncodeUint16Array(outputPtr, inputU16BePtr, strLength);
-    output.set(new Uint8Array(wm.memory.buffer, outputPtr, outputLength));
-    return outputLength;
+    const outputPtr = wm.utf8EncodeUint16Array(inputU16BePtr, strLength);
+    // the first 4 bytes is the outputByteLength in big-endian
+    const outputByteLength = new DataView(wm.memory.buffer, outputPtr, 4).getUint32(0);
+    output.set(new Uint8Array(wm.memory.buffer, outputPtr + 4, outputByteLength));
+    return outputByteLength;
   } finally {
     wm.free(inputU16BePtr);
     wm.free(outputPtr);
```

I guess calling `malloc()` is not so heavy.

Reply:
Hmm, in this case I recommend profile this better. Try manually measure (for example via console.time/console.timeEnd) for whole js `utf8EncodeUint16Array` method and for `wm.utf8EncodeUint16Array` call exclusively. Diff between times give you picture about which is price you pay for interop

Reply:
Tried to merge multiple malloc into a single one, but there were no effects.

Will try reference types https://github.com/WebAssembly/reference-types/blob/master/proposals/reference-types/Overview.md in a future.